### PR TITLE
authenticated users can join public teams

### LINKED ISSUE #80: Users can join public teams
For the case where a team is public and has `require_join_request=false`, users should be able to join the team via the team view page.

### PR DESCRIPTION
**Pull request by sethvincent**
closes #80 

This adds a button for authenticated users to the team view page letting them join the team if they are not already in it.

It also adds a `Sign in to join team` button if the user has not yet logged in.

All this is accomplished with an endpoint at `/api/teams/:id/join` that takes the user id of the current session and adds them to the team if allowed by the `team:join` permission. This is distinct from the endpoints for adding/updating members because it is assumed those are requested by team moderators. I think it makes sense for `join` to be separate.

If the user signs in with the `Sign in to join team` button it does not redirect them back to the team page. I think that would be an important thing to add. I am not yet sure how to accomplish that with this login flow, though. It might be best to make an issue for that and revisit it later.

### CODE REVIEW

**Comment by LanesGood** [comment]
@sethvincent this isn't working for me. My console logs
```
PUT http://localhost:8989/api/teams/20/join 401 (Unauthorized)
```

**Comment by LanesGood** [comment]
To add:
1. Need alert/callout when join is successful
2. Need to highlight/mark row with own user info
3. Need to highlight/mark group moderator row
4. Need to create Leave Team functionality, with “Are you sure” dialogue
5. Join Team button should be replaced with Leave Team button when member is not moderator

@kamicut @sethvincent should any of the above to-dos be included in this branch, or are these all separate issues?

File: pages/team.js [hunk 1]
```diff
@@ -181,6 +197,8 @@ export default class Team extends Component {
         <div className='page__heading'>
           <h2>{team.name}</h2>
           { isUserModerator ? <Button variant='primary' href={`/teams/${team.id}/edit`}>Edit Team</Button> : <div /> }
+          { userId && !isMember ? <Button variant='primary' onClick={() => this.joinTeam()}>Join Team</Button> : <div /> }
```
Comment (LanesGood) [inline pages/team.js]:
```suggestion
          { userId && !isMember ? <Button variant='primary' onClick={() => this.joinTeam()}>Join Team</Button> : '' }
```
Avoid rendering empty div, throwing off flex styling

File: pages/team.js [hunk 2]
```diff
@@ -181,6 +197,8 @@ export default class Team extends Component {
         <div className='page__heading'>
           <h2>{team.name}</h2>
           { isUserModerator ? <Button variant='primary' href={`/teams/${team.id}/edit`}>Edit Team</Button> : <div /> }
```
Comment (LanesGood) [inline pages/team.js]:
```suggestion
          { isUserModerator ? <Button variant='primary' href={`/teams/${team.id}/edit`}>Edit Team</Button> : ''}
```
Avoid rendering empty div, throwing off flex styling

**Comment by kamicut** [comment]
@LanesGood: 
1 is an enhancement
2 - 3 can be broken out into its own issue
4 - 5 this one is up to @sethvincent

File: pages/team.js [hunk 3]
```diff
@@ -181,6 +197,8 @@ export default class Team extends Component {
...
+          { userId && !isMember ? <Button variant='primary' onClick={() => this.joinTeam()}>Join Team</Button> : <div /> }
+          { !userId ? <Button variant='primary' href={`/login`}>Sign in to join team</Button> : <div /> }
```
Comment (LanesGood) [inline pages/team.js]:
```suggestion
          { !userId ? <Button variant='primary' href={`/login`}>Sign in to join team</Button> : '' }
```
Avoid rendering empty div, throwing off flex styling

File: pages/team.js [hunk 4]
```diff
@@ -180,7 +196,9 @@ export default class Team extends Component {
       <article className='inner page team'>
         <div className='page__heading'>
           <h2>{team.name}</h2>
-          { isUserModerator ? <Button variant='primary' href={`/teams/${team.id}/edit`}>Edit Team</Button> : <div /> }
+          { isUserModerator ? <Button variant='primary' href={`/teams/${team.id}/edit`}>Edit Team</Button> : ''}
+          { userId && !isMember ? <Button variant='primary' onClick={() => this.joinTeam()}>Join Team</Button> : <div /> }
```
Comment (sethvincent) [inline pages/team.js]:
```suggestion
          { userId && !isMember ? <Button variant='primary' onClick={() => this.joinTeam()}>Join Team</Button> : '' }
```

**Comment by sethvincent** [comment]
@LanesGood i'll add 4 & 5 in this pr.

1 we should tackle as a more comprehensive toast/status message pr. That's kinda what this issue is for: https://github.com/developmentseed/osm-teams/issues/49

**Comment by sethvincent** [comment]
Change of plans: we'll also do 4 & 5 as a separate pr. Issue for that here: https://github.com/developmentseed/osm-teams/issues/94

**Pull request merged**